### PR TITLE
Stabilize step_by by adding it to Iterator (issue #27741)

### PR DESCRIPTION
**Pull request by ivandardi**
Inspired by itertools' `take()` method. See issue #27741

### CODE REVIEW

**Comment by ivandardi** [comment]
This is my first commit to Rust! :D ~~I probably screwed something up~~

I may need help with the docstrings. I feel like they can be improved. Also, I couldn't get it to compile without adding the `#[stable]` tags. Which are the proper tags to put there?

File: src/libcore/iter/iterator.rs [hunk 1]
```diff
@@ -258,6 +258,32 @@ pub trait Iterator {
...
+    ///
+    /// assert_eq!(iter.next(), Some(&0));
+    /// assert_eq!(iter.next(), Some(&2));
+    /// assert_eq!(iter.next(), Some(&4));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
+    fn every_nth(self, step: usize) -> EveryNth<Self> where Self: Sized {
+        assert!(step != 0);
```
Comment (killercup) [inline src/libcore/iter/iterator.rs]: Not sure if I'd expect this method to panic, especially if I'm using it programmatically. I'm in no position to define this, but: If this assert stays, I'd add a custom message.

File: src/libcore/tests/iter.rs [hunk 2]
```diff
@@ -145,6 +145,22 @@ fn test_iterator_chain_find() {
 }
 
 #[test]
+fn test_every_nth_one() {
```
Comment (killercup) [inline src/libcore/tests/iter.rs]: Maybe add another test (with `#[should_panic]`) to test `.every_nth(0)`?

Comment (KalitaAlexey) [inline src/libcore/tests/iter.rs]: Yes, such a test should be added.

File: src/libcore/iter/iterator.rs (hunk 1 above)
Comment (nagisa) [inline src/libcore/iter/iterator.rs]: I would personally expect the behaviour of `.peek()` for `every_nth(0).next()`. Troublesome to implement though.

**Comment by ollie27** [comment]
How about if the number passed to `every_nth` is the number of elements that it should skip? It would mean `every_nth(0)` would return every element, `every_nth(1)` would return every other element and so on. It would match `nth` and remove the awkward `panic`.

File: src/libcore/iter/iterator.rs (hunk 1 above)
Comment (Stebalien) [inline src/libcore/iter/iterator.rs]: @nagisa more like impossible without cloning (peak returns a reference, `next` would return a value).

From the description ("skipping `step` elements at a time"), I'd expect `every_nth(0)` to return every item and `every_nth(1)` to skip every other item. This would also be consistent with `nth`.

Comment (ivandardi) [inline src/libcore/iter/iterator.rs]: Yeah, I think it would be best to have that behavior instead. I'll make the changes now.

**Comment by ivandardi** [comment]
@ollie27 How would `every_nth(1)` work?

Like this:
```rs
let mut it = (0..).every_nth(1).take(3);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(2));
assert_eq!(it.next(), Some(4));
assert_eq!(it.next(), None);
```

Or like this:
```rs
let mut it = (0..).every_nth(1).take(3);
assert_eq!(it.next(), Some(1));
assert_eq!(it.next(), Some(3));
assert_eq!(it.next(), Some(5));
assert_eq!(it.next(), None);
```

**Comment by ollie27** [comment]
@ivandardi that's a good question. Both seem reasonable to me.

**Comment by nagisa** [comment]
> How would every_nth(1) work?

Probably the first. You can get the behaviour of the 2nd code sample by skipping before hand, whereas you cannot get the first behaviour from the first.

**Comment by leonardo-m** [comment]
This is how Python behaves:

```
>>> range(10)[::0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: slice step cannot be zero
>>> range(10)[::1]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
>>> range(10)[::2]
[0, 2, 4, 6, 8]
```

**Comment by ivandardi** [comment]
Hm... It would be good to keep consistency with other languages too. If we keep the Python behavior, then the current implementation is correct and all it's needed is the `#[should_panic]` test.

After that we need to discuss whether or not to keep the name `every_nth` for clarity, how to fix the `#[stable]` tags and possibly the removal of the `Step` trait altogether.

**Comment by nagisa** [comment]
> If we keep the Python behavior, then the current implementation is correct

Too bad Python has a terrible behaviour in a lot of cases. I feel that at this point it is more in the garden of `T-libs` to decide what’s the optimal behaviour here.

Here’s what the documentation for the behaviour I would prefer:

``````
Returns an iterator adaptor which yields the next element and then skips the specified number of elements before yielding another element. `every_nth(0)` is an identity adaptor.

# Examples

```
let mut it = (0..).every_nth(0).take(2);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(1));
assert_eq!(it.next(), None);
```

```
let mut it = (0..).every_nth(2).take(3);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(3));
assert_eq!(it.next(), Some(6));
assert_eq!(it.next(), None);
```
``````

Implementation wise it also needs to be decided whether it is fine for iterator to be implemented as

```
fn next(&mut self) -> Option<Self::Item> { let n = self.next(); self.take(self.n).count(); n }
// as opposed to
fn next(&mut self) -> Option<Self::Item> { if self.first_take { self.next() } else { self.skip(self.n).next() } }
```

as these two have distinct behaviour in terms of the state the underlying iterator is after a `next`. Implementation approach will likely dictate the documentation wording as well.

**Comment by nagisa** [comment]
I’ll take liberty of tagging this as waiting-on-team, as there’s clearly nothing the author can act on.

File: src/libcore/iter/mod.rs [hunk 3]
```diff
@@ -520,6 +520,35 @@ impl<I> Iterator for Cycle<I> where I: Clone + Iterator {
...
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<I> Iterator for EveryNth<I> where I: Iterator {
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let elt = self.iter.next();
+        if self.step > 0 {
+            self.iter.nth(self.step - 1);
```
Comment (johncf) [inline src/libcore/iter/mod.rs]: If `nth` returns `None` here, then the behavior of calling `next` again is [unspecified](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next). Shouldn't that be avoided?

Comment (ivandardi) [inline src/libcore/iter/mod.rs]: Well, the iterator could be fused so that it never returns a `Some(T)` again. But I don't know if that's desired behavior, because one could voluntarily `.fuse()` beforehand.

Comment (johncf) [inline src/libcore/iter/mod.rs]: But the problem here is that `iter.next()` could return `Some(..)` and internally hit a `None`, but calling `next()` again could return `Some(..)`. So the user is not actually informed that the iterator returned `None` once, making the behavior of `every_nth` itself unspecified without using `fuse`.

Comment (ivandardi) [inline src/libcore/iter/mod.rs]: Itertools fuses the iterator internally too, so I don't see why not fuse it here too. We'd just need to leave it explicit in the documentation.

Comment (johncf) [inline src/libcore/iter/mod.rs]: I can think of a way to avoid fusing. But that would involve consuming an element when `every_nth` is called, which might be against the "laziness" principle of iterator combinators.

Comment (ranma42) [inline src/libcore/iter/mod.rs]: The second implementation shown by @nagisa avoids fusing and is as lazy as expected.

Comment (johncf) [inline src/libcore/iter/mod.rs]: @ranma42 But it has the same complexity as fusing (has a branch on every `next`). That branching can be avoided if first element is consumed in advance.

Comment (ranma42) [inline src/libcore/iter/mod.rs]: At least in the basic use case of stepping through ranges, the compiler should be able to inline the whole thing and get rid of every abstraction (the fuse iterator or the `first_take` flag, `nth` etc). For other use cases, it might be impossible to optimise away the branch, but it might also be very bad to consume elements in advance.

Comment (johncf) [inline src/libcore/iter/mod.rs]: > but it might also be very bad to consume elements in advance.

I wasn't clear enough, I think. I said consume "the first element" in advance, not "all." I'm not quite sure if even consuming the first element in advance could be "bad," except perhaps weakly breaking a principle.

Comment (ranma42) [inline src/libcore/iter/mod.rs]: Even if the fused/`first_take` approaches have the same branching, the current implementation and the one I pointed out do not have the same complexity from the point of view of work on the inner iterator, because of the early `nth`.
If the final consumer of the `every_nth` iterator only takes the first element, the underlying iterator will consume 1+n elements with the current implementation (as in @nagisa's first option), while one might expect it to only consume 1 element (as in the second option).

Comment (ranma42) [inline src/libcore/iter/mod.rs]: @johncf I think a similar rationale applies to [`Peekable`](https://doc.rust-lang.org/std/iter/struct.Peekable.html). It would be easier to write it so that it always contains the current element (or `None`). Instead it was explicitly coded to keep track of whether the element has been pulled from the inner iterator.

Comment (johncf) [inline src/libcore/iter/mod.rs]: @ranma42 I'm guessing, that rationale would indeed matter when the iterator is consuming data from a network stream, for instance. In that case, consuming any element in advance might cause unexpected blocking behavior.

File: src/libcore/iter/iterator.rs [hunk 4]
```diff
@@ -258,6 +258,32 @@ pub trait Iterator {
...
+    ///
+    /// let mut iter = a.iter().every_nth(2);
+    ///
+    /// assert_eq!(iter.next(), Some(&0));
+    /// assert_eq!(iter.next(), Some(&2));
+    /// assert_eq!(iter.next(), Some(&4));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
```
Comment (BurntSushi) [inline src/libcore/iter/iterator.rs]: You'll want to fix this attribute. It should be given its own feature name and it should be unstable.

**Comment by BurntSushi** [comment]
cc @rust-lang/libs Are we sufficiently past the `step_by` stuff that we'd be willing to take `every_nth`?

File: src/libcore/iter/iterator.rs (hunk 4 above)
Comment (ivandardi) [inline src/libcore/iter/iterator.rs]: Yes, I plan to do that. I just didn't know how when I made the pull request.  Also, the feature name depends on whether we get rid of the `Step` trait or if we'll keep the step trait and have this as `every_nth`. I personally prefer the former, since there's some code that already uses `step_by` and wouldn't break as bad with a change in the implementation. That would mean, however, that I'd have to reimplement `Iterator for Range`, which is not hard, but would definitely take some time. Once all that gets figured out, I'll put the proper attributes.

Comment (scottmcm) [inline src/libcore/iter/iterator.rs]: Does taking the name `step_by` really require _removing_ the `Step` trait?

Removing `step`, `steps_between`, and `is_negative` from `Step` seems logical, as is replacing `core::iter::StepBy` with the logic worked out here.  But redoing all the `Range` iterators feels like it shouldn't be necessary for this change, and I don't see how the existence of `Step` conflicts with it.

Comment (ivandardi) [inline src/libcore/iter/iterator.rs]: Because if we rename `every_nth` to `step_by`, then there's going to be a method name conflict for ranges, because there would be a `step_by` from the Iterator trait, and a [`step_by` from the Range struct](https://doc.rust-lang.org/std/ops/struct.Range.html#method.step_by).

Technically we wouldn't need to remove the `Step` trait, but if we switch the `Range::step_by` implementation to the iterator version, then there's going to be no use for the `Step` trait in the core lib. Besides, it looks like the only implementors of `Step` are the normal integers. Nobody really uses the `Step` trait to increment an integer when they can just use `+= 1`. And if people really wanted it, we could just implement `Iterator` for them and have them return the integer plus one. It would work the same.

Comment (BurntSushi) [inline src/libcore/iter/iterator.rs]: @ivandardi I think we should just go with a separate feature name for now. We can do the removal in another PR.

**Comment by BurntSushi** [comment]
@ivandardi The libs team discussed this today, and it seems like we're all on board with this direction. It looks like there are still some changes you wanted to make before merging though? For example, is it intentional that `every_nth(0)` panics?

**Comment by nagisa** [comment]
@BurntSushi I feel like there has been some mis-comunication? It was on the libs teams to decide on one of the many proposed behaviours, and I do not see any decision wrt that.

**Comment by BurntSushi** [comment]
@nagisa I see. I'll take a closer look and try to summarize.

**Comment by ivandardi** [comment]
Yes, it's expected that `every_nth(0)` panics, otherwise it would enter an infinite loop of yielding the first element forever. It's consistent with the behavior seen in Python too.

I'll submit a commit making the changes to the feature name and making it unstable hopefully later today.

**Comment by nagisa** [comment]
> Yes, it's expected that every_nth(0) panics, otherwise it would enter an infinite loop of yielding the first element forever. It's consistent with the behavior seen in Python too.

You did agree before that having `every_nth(0)` be identity iterator transformer could make sense. You don’t think so anymore?

**Comment by ivandardi** [comment]
@nagisa Nope. Makes more sense for `every_nth(1)` to be the identity, because it means that it will give every 1st element, aka the next element. It also goes in hand with what other languages have, and it's important to have familiar behavior. Yeah, I know I went back and forth in this, but this hasn't been easy to decide! XD But I think everyone is pretty settled on it now.

Also, I was testing with `step_by(0)` always yielding the first element of the iterable and I actually accidentally crashed my browser a lot of times in the playground while testing it out. If you have a `for i in (0..10).every_nth(0)`, it will loop forever unless you have a break inside the loop. Combine that with the fact that people might have something like

```rs
usize step = do_calculation()
for i in (0..len).step_by(step)
```
where `step` might accidentally be `0` and it's a really bad thing. I now see why Python throws an exception when the step parameter in their ranges is 0 and that's why I'm also making it panic on `step_by(0)`.

**Comment by ollie27** [comment]
> Makes more sense for every_nth(1) to be the identity, because it means that it will give every 1st element, aka the next element.

That doesn't match [`nth`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.nth):

> Like most indexing operations, the count starts from zero, so nth(0) returns the first value, nth(1) the second, and so on.

**Comment by ivandardi** [comment]
That's why I think it's best to rename this to `step_by`. That way it doesn't make any correlation with `nth`.

**Comment by scottmcm** [comment]
@ivandardi It looks like GitHub is showing a ton of unrelated changes in here now.  Does it need to be rebased?

File: src/libcore/iter/iterator.rs [hunk 5]
```diff
@@ -258,6 +258,43 @@ pub trait Iterator {
         None
     }
 
+    /// Returns an iterator adaptor that goes over the original iterator,
+    /// skipping `step` elements at a time.
```
Comment (BurntSushi) [inline src/libcore/iter/iterator.rs]: I'm struggling with these docs. It says that it *skips* `step` elements at a time, but in the example, `step == 2` but it's only skipping one element at a time.

**Comment by BurntSushi** [comment]
> Implementation wise it also needs to be decided whether it is fine for iterator to be implemented as

@nagisa Could you say more about how a caller could witness a difference in behavior between your two proposed implementation?

**Comment by BurntSushi** [comment]
cc @rust-lang/libs See brief summary of choices below:

OK, so I did a more careful review and it looks like we have two strategies. The first is a variant of what Python does (roughly copied from this PR's unit test):

```rust
let mut it = (0..).every_nth(0).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), None);

let mut it = (0..).every_nth(1).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(1));
assert_eq!(it.next(), Some(2));
assert_eq!(it.next(), Some(3));
assert_eq!(it.next(), None);

let mut it = (0..).every_nth(3).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(3));
assert_eq!(it.next(), Some(6));
assert_eq!(it.next(), Some(9));
assert_eq!(it.next(), None);
```

In this case, `every_nth(1)` is the identity: it yields the underlying iterator unchanged. The interesting case in this example is that `every_nth(0)` just returns the first element and stops, which is a bit strange. This PR originally panicked in that situation, which is analogous to what Python does if you try, e.g., `xs[::0]`.

The other strategy (as pointed out by @nagisa and @Stebalien) is to make `every_nth(0)` the identity iterator:

```rust
let mut it = (0..).every_nth(0).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(1));
assert_eq!(it.next(), Some(2));
assert_eq!(it.next(), Some(3));
assert_eq!(it.next(), None);

let mut it = (0..).every_nth(1).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(2));
assert_eq!(it.next(), Some(4));
assert_eq!(it.next(), Some(6));
assert_eq!(it.next(), None);

let mut it = (0..).every_nth(3).take(4);
assert_eq!(it.next(), Some(0));
assert_eq!(it.next(), Some(4));
assert_eq!(it.next(), Some(8));
assert_eq!(it.next(), Some(12));
assert_eq!(it.next(), None);
```

I personally think I prefer the latter, even though it is different from what Python does. It seems like there are fewer footguns, and it can be easily explained as "skips `N` elements." The Python approach is "skips `N-1` elements."

**Comment by nagisa** [comment]
> Could you say more about how a caller could witness a difference in behavior between your two proposed implementation?

Sure. So, imagine code as such:

```rust
fn foo(i: &mut Iterator<Item=i32>) {
    let d: Vec<_> = i.every_nth(2).take(2).collect(); // assuming `every_nth(0)` is `id`
}

fn main() {
    let itr = 0..10;
    foo(&mut itr);
    println!("{:?}", itr.next());
}
```

Now, there’s two distinct values the code could print, both potentially valid, depending on the semantics we decide on: `4` and `6`. 

`6` would be printed in case of eager implementation. That is:

* every_nth consumes 3 values from the iterator and yields the first of the three: `0`;
* every_nth consumes 3 values from the iterator and yeilds the first of the three: `3`;
* `itr.next()` returns `6`.

`4` would be printed in case of non-eager implementation. That is:

* `every_nth` checks a flag, since it is set, unsets the flag and yields the first value `0`;
* `every_nth` checks a flag, since it is unset, skips 2 values and yields the next one: `3`;
* `itr.next()` returns `4`.

**Comment by CryZe** [comment]
Okay, so the initial goal of this was to stabilized `step_by` by implementing it as a method on iterators. Someone then suggested that we could name it `every_nth` so it sounds similar to the `nth` method that already exists for Iterators. However during the implementation it turned out that naming it `every_nth` has two possible different semantics, depending on whether you think of it as an `index` or as a `delta`. So it would probably be better to go back to the original naming, which is not ambiguous. After some chatting with the author, it turned out that he wasn't aware that you can have a method impl on a struct and a trait impl with the same name, which was one of the reasons for not using `step_by`. 

Also I'd like to remind everyone that the actual use case we are trying to solve here, is getting `for (int i = start; i < end; i += delta)` into Rust. So this is supposed to be heavily used in for loops. With `step_by` you can obviously still cause the `+= 0` issue. A user could easily cause your program to run an infinite loop then (in case we opt into these semantics). However, if you think about it, this is a division by zero. You are dividing your whole range into slices of width 0, this isn't even defined mathematically. So in practice you will always have to handle this case as some special case. People will in practice use this for the for loops with the `+= delta` semantics, this is going to be the main use case. Now we can throw people off by removing the `0 from our numerical domain`, by saying that the user needs to `.step_by(49)` to replicate what they want to implement, `+= 50`. However this will not only cause tons of bugs, this doesn't even get rid of the division by 0. Because if this is actually indeed a user input, the user input will always be in terms of `+=`, as any other behavior would be even less expected for a common user of your application, than it would be for a programmer who is used to indexing from 0. So the user will input 50, and you would have to subtract 1 from the user input to pass it to `every_nth`. If the user used 0 as his input, you are going to subtract 1 from his value and underflow your usize. In debug, this will cause a panic, so even `every_nth` semantics would cause panics in practice. But even worse, in Release Mode the underflow wouldn't cause a panic, but instead skip your whole loop. 

So in practice, the `every_nth` behavior is unexpected to everyone trying to use it as `+=`, which is going to be main use case, will cause panics in debug mode and will skip your loop in release mode, which is very inconsistent behavior. So I suggest that we go back to the original naming which doesn't have any of these problems.

File: src/libcore/iter/iterator.rs (hunk 5 above)
Comment (ivandardi) [inline src/libcore/iter/iterator.rs]: Oh, I forgot to change that. I'll update the documentation soon. Thanks for reminding me!

**Comment by ivandardi** [comment]
Ok, finally managed to fix the derp up I did with the last commit. Note to self: don't git late at night.

In terms of backwards compatibility about the latest changes, people already have been using the `Range::step_by` unstable method. It behaves like this: `(0..10).step_by(3) == [0, 3, 6, 9]`. Even though it's experimental, people already have an expectation of how it should behave, and ultimately I think it's best to continue with this behavior, especially because if the current implementation is already like this, then there must have been a reason for it. On top of that, it doesn't break existing code, which is a plus!

Also, I provided a test implementation to yield the first element only once on `step_by(0)`. Ultimately, it was decided that it was inconsistent with the expected behavior, and that it should indeed panic on 0. Once this gets actually stabilized, then it could be implemented as a Clippy lint to prevent mistakes at compile time.

The current implementation takes in the laziness concern: it only skips elements on the actual call to `.next()`. On the question of complexity, the branching in the implementation should be fine on most CPUs, either because branch prediction will take place and always predict correctly, or even LLVM may be smart enough to generate bytecode without any branching.

To be able to rename it to `step_by`, I had to re-export `range::StepBy` as `range::DeprecatedStepBy`. I don't know if it would be ideal to remove it in this PR, so I'm leaving it open. Also, as a consequence of that, I had to temporarily use UFCS on the tests, because they use ranges and it was calling `Range::step_by` instead. Once `Range::step_by` gets removed, the tests can be updated to use the better syntax. I added a comment on them as a reminder. (Also, turns out that it had documentation that explains what it did already, so I just copied it over.)

File: src/libcore/iter/iterator.rs [hunk 6]
```diff
@@ -258,6 +258,40 @@ pub trait Iterator {
...
+    /// Creates an iterator starting at the same point, but stepping by
+    /// the given amount at each iteration.
+    ///
+    /// Note that it will always return the first element of the range,
+    /// regardless of the step given.
+    ///
+    /// # Panics
+    ///
+    /// If the given step is `0`, the method will panic if debug assertions are
+    /// enabled.
```
Comment (scottmcm) [inline src/libcore/iter/iterator.rs]: nit: This comment makes me think `debug_assert!`, but it's not.  Does it need more than "This method will panic if the given `step` is `0`."?

Comment (ivandardi) [inline src/libcore/iter/iterator.rs]: I just looked at how the other methods in `Iterator` that panic were phrased and tried doing something similar for consistency. 

I just checked now, and I didn't know of `debug_assert!`, which is what the other methods are using. I've been mislead! :P I'll fix it soon. I also have to figure out what to write on the Unstable Book to make that tidy check pass 🙄

**Comment by alexcrichton** [comment]
ping @BurntSushi looks like there's been a few updates, mind taking a look?

**Comment by BurntSushi** [comment]
@CryZe Great points, I think you've convinced me. This looks good to go. Mucho thanks to @ivandardi for your patience with this. :-) The failure looks spurious, so @bors r+

**Comment by ivandardi** [comment]
Awesome! Yeah, it took one month to get this going! At least I'm glad it works. After this, there needs to be another PR to take care of `StepByDeprecated` and reimplement `Iterator` for `Range`s.

Oh, and a last question: Is it my Github name that gets added to the contributors list automatically? If yes, then I just updated it. :D

**Comment by BurntSushi** [comment]
> Oh, and a last question: Is it my Github name that gets added to the contributors list automatically? If yes, then I just updated it. :D

Hmm not sure. @steveklabnik ?

**Comment by CryZe** [comment]
I'm pretty sure it's just the name and email you committed as.

**Comment by alexcrichton** [comment]
@bors: retry

* [appveyor network issues](https://appveyor.statuspage.io/incidents/2g5jnq5cy1g1) #40474

**Comment by birkenfeld** [comment]
Isn't there a specialization for the `Range` types missing?

**Comment by quininer** [comment]
After I switch from `step_by` to `iterator_step_by`, the bench shows a serious performance degradation. https://github.com/quininer/newhope/commit/038aab77bbeca8e9273c375e59c471904c43ea78

before:
```
running 3 tests
test bench_newhope_keygen  ... bench:     117,488 ns/iter (+/- 21,484)
test bench_newhope_shareda ... bench:      49,293 ns/iter (+/- 17,389)
test bench_newhope_sharedb ... bench:     220,014 ns/iter (+/- 168,978)
```

after:
```
running 3 tests
test bench_newhope_keygen  ... bench:   1,189,949 ns/iter (+/- 353,381)
test bench_newhope_shareda ... bench:     542,612 ns/iter (+/- 80,311)
test bench_newhope_sharedb ... bench:   1,609,396 ns/iter (+/- 420,654)
```
